Source/NSLayoutConstraint.m

```diff
@@ -0,0 +1,569 @@
+    }
+  else if ([@"Baseline" isEqualToString: str])
+    {
+      a = NSLayoutAttributeBaseline;
+    }  
+  else if ([@"FirstBaseline" isEqualToString: str])
+    {
+      a = NSLayoutAttributeFirstBaseline;
+    }
+  else if ([@"NotAnAttribute" isEqualToString: str])
+    {
+      a = NSLayoutAttributeNotAnAttribute;
+    }
+
+  return a;
+}
+
++ (NSString *) _relationToString: (NSLayoutRelation)rel
+{
+  NSString *relation = nil;
+
+  switch (rel)
+    {
+    case NSLayoutRelationLessThanOrEqual:
+      relation = @"<=";
+      break;
+    case NSLayoutRelationEqual:
+      relation = @"=";
+      break;
+    case NSLayoutRelationGreaterThanOrEqual:
+      relation = @">=";
+      break;
+    default:
+      break;
+    }
+
+  return relation;
+}
+
++ (NSLayoutRelation) _stringToRelation: (NSString *)str
+{
+  NSLayoutRelation r = 0;
+
+  if ([@"<=" isEqualToString: str])
+    {
+      r = NSLayoutRelationLessThanOrEqual;
+    }
+  else if ([@"=" isEqualToString: str])
+    {
+      r = NSLayoutRelationEqual;
+    }
+  else if ([@">=" isEqualToString: str])
+    {
+      r = NSLayoutRelationGreaterThanOrEqual;
+    }
+  
+  return r;
+}
+
++ (void) _activateConstraint: (NSLayoutConstraint *)constraint
+{
+  [activeConstraints addObject: constraint];
+  activeConstraints = [[activeConstraints sortedArrayUsingSelector: @selector(compare:)] mutableCopy];
+}
+
++ (void) _removeConstraint: (NSLayoutConstraint *)constraint
+{
+  [activeConstraints removeObject: constraint];
+}
+
++ (NSArray *) constraintsWithVisualFormat: (NSString *)fmt 
+                                  options: (NSLayoutFormatOptions)opt 
+                                  metrics: (NSDictionary *)metrics 
+                                    views: (NSDictionary *)views
+{
+  NSMutableArray *array = [NSMutableArray arrayWithCapacity: 10];
+  return array;
+}
+
+// Designated initializer...
++ (instancetype) constraintWithItem: (id)view1 
+                          attribute: (NSLayoutAttribute)attr1 
+                          relatedBy: (NSLayoutRelation)relation 
+                             toItem: (id)view2 
+                          attribute: (NSLayoutAttribute)attr2 
+                         multiplier: (CGFloat)mult 
+                           constant: (CGFloat)c
+{
+  NSLayoutConstraint *constraint =
+    [[NSLayoutConstraint alloc] initWithItem: view1
+                                   attribute: attr1
+                                   relatedBy: relation
+                                      toItem: view2
+                                   attribute: attr2
+                                  multiplier: mult
+                                    constant: c];
+
+  AUTORELEASE(constraint);
+  return constraint;
+}
+
++ (void) activateConstraints: (NSArray *)constraints
+{
+  NSEnumerator *en = [constraints objectEnumerator];
+  NSLayoutConstraint *c = nil;
+
+  while ((c = [en nextObject]) != nil)
+    {
+      [NSLayoutConstraint _activateConstraint: c];
```

Comment:
Looks like you do the sorting way to often here.

Reply:
I had initially implemented this without doing any sorting... just an insertion sort when an object is added.   I am going to return to this approach so it's not sorting it every time.

Reply:
I think this should be deferred since we are discussing not applying the constraints.

Reply:
I do not see how this argument is related. Just copy the code from _activateConstraint: to here and move the sorting out of the loop.